deduction_rules.py

```diff
@@ -112,11 +112,14 @@ def search_and_ban_in_subset(cells_to_check):
         allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
         pairs = search_hidden_pairs(allowed_numbers)
         for pair,except_nums in pairs.items():
-            current_hidden_cell0 = cells_to_check[pair[0]]
-            current_hidden_cell1 = cells_to_check[pair[1]]
-            cells_used = sudoku.allowed[current_hidden_cell0[0]][current_hidden_cell0[1]].notNones()+sudoku.allowed[current_hidden_cell1[0]][current_hidden_cell1[1]].notNones()
-            made_deduction |= _ban_numbers(sudoku, current_hidden_cell0, filter(lambda x: x not in except_nums, allowed_numbers[pair[0]]), "hidden_pair",cells_used)
-            made_deduction |= _ban_numbers(sudoku, current_hidden_cell1, filter(lambda x: x not in except_nums, allowed_numbers[pair[1]]), "hidden_pair",cells_used)
+            cell0 = cells_to_check[pair[0]]
+            cell1 = cells_to_check[pair[1]]
+            cells_used = []
+            for cell in cells_to_check:
+                if cell not in (cell0, cell1):
+                    cells_used += sudoku.allowed[cell[0]][cell[1]].notNones()
```

Comment:
This is technically correct, but there is more information listed here than necessary: in both `hidden_pair` and `hidden_triple` we only care about the *"hidden"* numbers. For example, imagine that 3 and 4 can only be written to the first two cells of a row. For a `hidden_pair` deduction we use the information that 3 and 4 CAN'T be written to any other cell of the row, but don't use the fact that 5 can't be written the the fourth cell or that 6 can't be written to the last.

To achieve proper k-optimization, additional filtering is necessary here (and in `hidden_trios`). Something like (this is incorrect code!):
```
used_nums = # numbers we apply the rule for (1-9) #
...
for ...
    if ...
        cells_used += [sudoku.allowed[cell[0]][cell[1]][n] for n in used_nums if sudoku.allowed[cell[0]][cell[1]][n] is not None]
```
Note that we *would* be able to omit the `if` check in the list comprehension (as we already know these numbers cannot be written to the other cells in this row/col/sec) if you __uncommented lines 139-140 in `sudoku.py`__.  Please do so, as we also have to save if these numbers cannot be written to other cells *because* these cells are already filled (and the code above fails to do that). Without uncommenting those lines, it would be quite difficult to write this part od the code.

Reply:
Oh, I'm sorry, I hope it is fixed now.
I uncommented it earlier, but I didnt pushed it. But I don't use it now nowhere. And I wrote a helper function for quering the allowed number. I use it everywhere rather than using a loong expression (`sudoku.allowed[index[0]][index[1].allowed()`) that I always forget.